### DISCUSSION TITLE
How to hide a window on MacOS?

### DISCUSSION THREAD

**Discussion by manfromarce** [AvaloniaUI/Avalonia#7961] [Q&A]
On MacOS, users expect the window to be hidden but not closed when clicking on the X button.
So I've added this code to the Window.Closing event:

```
if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX) && !ExplicitCloseWindowRequested)
            {
                     e.Cancel = true;
                     this.Hide();
            }
```
The problem is that the window is no longer shown when clicking on the Dock icon.
Also, I would like that when the window is hidden the focus goes to another app or another of my app's windows (instead of staying on the hidden window, causing the global menu to show only the App submenu).
I've also added a MonoMac dependecy and tried different approaches but none of them worked:
`this.PlatformImpl.Hide(); //same as this.Hide()`
```
var nativeWindow = NSWindow.FromWindowRef(this.PlatformImpl.Handle.Handle);
nativeWindow.BeginInvokeOnMainThread(() => nativeWindow.OrderOut(nativeWindow));
//this gives a thread exception - same effect as nativeWindow.OrderOut(nativeWindow))
```
`NSApplication.SharedApplication.Hide(nativeWindow); //this closes the window altogether (?)` 

Setting WindowState to Minimized works but I don't want the additional window icon to stay on the taskbar.

Has anyone ever managed to achieve this?

**Comment by duke-darius** [comment]
I'd be very intrigued to here an answer on this

**Comment by carina-studio** [comment]
You need to create and set your own [**NSApplicationDelegate**](https://developer.apple.com/documentation/appkit/nsapplicationdelegate) object to [**NSApplication.delegate**](https://developer.apple.com/documentation/appkit/nsapplication/1428705-delegate). And implement method to handle [**applicationShouldHandleReopen(_:hasVisibleWindows:)**](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428638-applicationshouldhandlereopen).

Please noted that you will replace the original **NSApplicationDelegate** set by Avalonia, so you need to  keep the original **NSApplicationDelegate** instance and call back to it in your own implementation. You can refer to [app.mm](https://github.com/AvaloniaUI/Avalonia/blob/master/native/Avalonia.Native/src/OSX/app.mm) to check which methods are originally implemented by Avalonia so that you can just call back to these methods.

[Here](https://github.com/carina-studio/AppSuiteBase/blob/fe7729ae0c13d373ebc847e42a429648993bd082/Core/AppSuiteApplication.MacOS.cs#L40) is my implementation in base library of my applications, hope it can help you.

**Comment by MonsieurTib** [comment]
Is there a more straightforward approach to accomplish this task?

**Comment by MammadovAdil** [comment]
After looking into Avalonia source code, I found a better way of doing it using `IActivatableLifetime`. Your `MainWindow.xaml.cs` file should look something _similar_ to below. Please note that it's NOT complete code for hiding, I emphasized mainly showing when clicking on the dock icon:

``` csharp
public class MainWindow : Window
{
    public MainWindow()
    {
        InitializeComponent();
    }

    private void InitializeComponent()
    {
        AvaloniaXamlLoader.Load(this);
        SubscribeToActivatableApplicationEvents();
    }


    protected override void OnClosing(WindowClosingEventArgs e)
    {
        if (Application.Current!.ApplicationLifetime is
            IClassicDesktopStyleApplicationLifetime { ShutdownMode: ShutdownMode.OnExplicitShutdown })
        {
            if (e.CloseReason is not WindowCloseReason.ApplicationShutdown)
            {
                // hide your window here, also consider case when window is in fullscreen mode
                e.Cancel = true;
            }

            // Unsubscribe from application events only when the app is shutting down
            if (e.CloseReason is WindowCloseReason.ApplicationShutdown or WindowCloseReason.OSShutdown)
                UnsubscribeFromActivatableApplicationEvents();
        }

        base.OnClosing(e);
    }

    private void SubscribeToActivatableApplicationEvents()
    {
        if (!TryGetActivatableLifetime(out var activatableLifetime))
            return;

        activatableLifetime.Activated += OnApplicationActivated;
    }

    private void UnsubscribeFromActivatableApplicationEvents()
    {
        if (!TryGetActivatableLifetime(out var activatableLifetime))
            return;

        activatableLifetime.Activated -= OnApplicationActivated;
    }

    private void OnApplicationActivated(object sender, EventArgs e)
    {
        if (!IsVisible)
            Show();
    }

    private static bool TryGetActivatableLifetime(out IActivatableLifetime activatableLifetime)
    {
        if (Application.Current is null ||
            !Application.Current.TryGetFeature<IActivatableLifetime>(out var foundActivatableLifetime))
        {
            activatableLifetime = null;
            return false;
        }

        activatableLifetime = foundActivatableLifetime;
        return true;
    }
}
```
